arch/arm64/src/imx9/imx9_tpm_pwm.c

```diff
@@ -0,0 +1,686 @@
+    {
+      /* Handle channel specific setup */
+
+      for (i = 0; i < CONFIG_PWM_NCHANNELS; i++)
+        {
+          if (ret != OK || info->channels[i].channel == -1)
+            {
+              break;
+            }
+
+          pwm_update_duty(priv, info->channels[i].channel,
+                          info->channels[i].duty);
+        }
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: pwm_stop
+ *
+ * Description:
+ *   Stop the pulsed output and reset the timer resources
+ *
+ * Input Parameters:
+ *   dev - A reference to the lower half PWM driver state structure
+ *
+ * Returned Value:
+ *   Zero on success; a negated errno value on failure
+ *
+ * Assumptions:
+ *   This function is called to stop the pulsed output at anytime.
+ *
+ ****************************************************************************/
+
+static int pwm_stop(struct pwm_lowerhalf_s *dev)
+{
+  struct imx9_pwmtimer_s *priv = (struct imx9_pwmtimer_s *)dev;
+  int i;
+
+  pwminfo("PWM%d stop\n", priv->id);
+
+  /* Check that timer is valid */
+
+  if (priv == NULL)
+    {
+      return -EINVAL;
+    }
+
+  /* Disable the channels */
+
+  for (i = 0; i < priv->n_channels; i++)
+    {
+      pwm_putreg(priv, IMX9_TPM_CXSC_OFFSET(CHMUX(priv, i)), 0);
+    }
+
+  /* Set frequency to 0 */
+
+  pwm_update_frequency(priv, 0);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: pwm_ioctl
+ *
+ * Description:
+ *   Lower-half logic may support platform-specific ioctl commands
+ *
+ * Input Parameters:
+ *   dev - A reference to the lower half PWM driver state structure
+ *   cmd - The ioctl command
+ *   arg - The argument accompanying the ioctl command
+ *
+ * Returned Value:
+ *   Zero on success; a negated errno value on failure
+ *
+ ****************************************************************************/
+
+static int pwm_ioctl(struct pwm_lowerhalf_s *dev, int cmd,
+                     unsigned long arg)
+{
+  return -ENOTTY;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: imx9_tpm_pwm_init
+ *
+ * Description:
+ *   Initialize tpm blocks to generate EPWM.
+ *
+ * Input Parameters:
+ *   pwmid - A number identifying the pwm block. The number of valid
+ *           IDs varies depending on the configuration.
+ *
+ * Returned Value:
+ *   On success, a pointer to the lower half PWM driver is
+ *   returned. NULL is returned on any failure.
+ *
+ ****************************************************************************/
+
+struct pwm_lowerhalf_s *imx9_tpm_pwm_init(tpm_pwm_id_t pwmid)
+{
+  struct imx9_pwmtimer_s *lower = NULL;
+  int i;
+
+  for (i = 0; i < sizeof(g_pwmdev) / sizeof(struct imx9_pwmtimer_s); i++)
```

Comment:
Hm, afaik nitems is BSD & Linux, but not standard. I changed it to sizeof(g_pwmdev) / sizeof(g_pwmdev[0]), which looks better to my eye than the orginal form. I can use nitems as well if you insist, though ..

Reply:
No push to use it. I was thinking that since code is in kernel and `nitems` is defined in Nuttx headers then it is safe to use it. We can keep `sizeof`